Unable to use rp2040_pio_led_strip building on Linux x86_64

### The problem

When updating to ESPHome 2023.6.0 I saw "Rp2040 pio ledstrip esphome#4818" listed as a new integration. When trying it I get this error:
`INFO ESPHome 2023.6.0
INFO Reading configuration /config/esphome/cubelight.yaml...
INFO Generating C++ source...
INFO Generating PIO assembly code
INFO Assembling PIO assembly code
Tool Manager: Installing earlephilhower/tool-pioasm-rp2040-earlephilhower
INFO Installing earlephilhower/tool-pioasm-rp2040-earlephilhower
Error: Could not find the package with 'earlephilhower/tool-pioasm-rp2040-earlephilhower' requirements for your system 'linux_x86_64'
ERROR PIO assembly failed`

Maybe related but I found issue #4541, which suggests its broken for all environments? If so then I suggest the integration isn't listed in the release notes.

### Which version of ESPHome has the issue?

2023.6.0

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

_No response_

### What platform are you using?

RP2040

### Board

PicoW

### Component causing the issue

rp2040_pio_led_strip

### Example YAML snippet

```yaml
light:
  - platform: rp2040_pio_led_strip
    name: cube_light
    id: cube_light
    pin: GPIO15
    num_leds: 50
    pio: 0
    rgb_order: GRB
    chipset: WS2812B
```


### Anything in the logs that might be useful for us?

```txt
INFO ESPHome 2023.6.0
INFO Reading configuration /config/esphome/cubelight.yaml...
INFO Generating C++ source...
INFO Generating PIO assembly code
INFO Assembling PIO assembly code
Tool Manager: Installing earlephilhower/tool-pioasm-rp2040-earlephilhower
INFO Installing earlephilhower/tool-pioasm-rp2040-earlephilhower
Error: Could not find the package with 'earlephilhower/tool-pioasm-rp2040-earlephilhower' requirements for your system 'linux_x86_64'
ERROR PIO assembly failed
```


### Additional information

_No response_

@ssieb that gets it much further but it then fails at:
```
src/main.cpp: In function 'void setup()':
src/main.cpp:169:67: error: 'rp2040_pio_led_strip_pixels_program' was not declared in this scope
  169 |   rp2040_pio_led_strip_rp2040pioledstriplightoutput->set_program(&rp2040_pio_led_strip_pixels_program);
      |                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/main.cpp:170:72: error: 'rp2040_pio_led_strip_driver_pixels_init' was not declared in this scope
  170 |   rp2040_pio_led_strip_rp2040pioledstriplightoutput->set_init_function(rp2040_pio_led_strip_driver_pixels_init);
      |                                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Archiving /data/cubelight/.pioenvs/cubelight/lib1c6/liblwIP_CYW43.a
Compiling /data/cubelight/.pioenvs/cubelight/libce7/lwIP_Ethernet/LwipIntf.cpp.o
Compiling /data/cubelight/.pioenvs/cubelight/libce7/lwIP_Ethernet/LwipIntfCB.cpp.o
*** [/data/cubelight/.pioenvs/cubelight/src/main.cpp.o] Error 1
========================== [FAILED] Took 5.07 seconds ==========================
```

--- a/esphome/components/rp2040/__init__.py
+++ b/esphome/components/rp2040/__init__.py
@@ -16,8 +16,7 @@ from esphome.const import (
     KEY_TARGET_PLATFORM,
 )
 from esphome.core import CORE, coroutine_with_priority, EsphomeError
-from esphome.helpers import mkdir_p, write_file
-import esphome.platformio_api as api
+from esphome.helpers import mkdir_p, write_file, copy_file_if_changed
 
 from .const import KEY_BOARD, KEY_PIO_FILES, KEY_RP2040, rp2040_ns
 
@@ -193,25 +192,19 @@ def generate_pio_files() -> bool:
         pio_path = CORE.relative_build_path(f"src/pio/{key}.pio")
         mkdir_p(os.path.dirname(pio_path))
         write_file(pio_path, data)
-        _LOGGER.info("Assembling PIO assembly code")
-        retval = api.run_platformio_cli(
-            "pkg",
-            "exec",
-            "--package",
-            "earlephilhower/tool-pioasm-rp2040-earlephilhower",
-            "--",
-            "pioasm",
-            pio_path,
-            pio_path + ".h",
-        )
-        includes.append(f"pio/{key}.pio.h")
-        if retval != 0:
-            raise EsphomeError("PIO assembly failed")
 
     write_file(
         CORE.relative_build_path("src/pio_includes.h"),
         "#pragma once\n" + "\n".join([f'#include "{include}"' for include in includes]),
     )
+
+    dir = os.path.dirname(__file__)
+    build_pio_file = os.path.join(dir, "build_pio.py.script")
+    copy_file_if_changed(
+        build_pio_file,
+        CORE.relative_build_path("build_pio.py"),
+    )
+
     return True
 
 

--- a/esphome/components/rp2040/__init__.py
+++ b/esphome/components/rp2040/__init__.py
@@ -192,6 +192,7 @@ def generate_pio_files() -> bool:
         pio_path = CORE.relative_build_path(f"src/pio/{key}.pio")
         mkdir_p(os.path.dirname(pio_path))
         write_file(pio_path, data)
+        includes.append(f"pio/{key}.pio.h")
 
     write_file(
         CORE.relative_build_path("src/pio_includes.h"),

--- a/esphome/components/rp2040/build_pio.py.script
+++ b/esphome/components/rp2040/build_pio.py.script
@@ -42,6 +42,6 @@ if PIO_FILES:
     for filename in PIO_FILES:
         if verbose:
             print(f"  {filename}")
-        subprocess.run([PIOASM_EXE, "-o", "c-sdk", filename, "{filename}.h"])
+        subprocess.run([PIOASM_EXE, "-o", "c-sdk", filename, f"{filename}.h"])
     if verbose:
         print("==============================================")

--- a/esphome/components/rp2040/build_pio.py.script
+++ b/esphome/components/rp2040/build_pio.py.script
@@ -1,0 +1,47 @@
+"""
+Custom pioasm compiler script for platformio.
+(c) 2022 by P.Z.
+
+Sourced 2023/06/23 from https://gist.github.com/hexeguitar/f4533bc697c956ac1245b6843e2ef438
+
+Modified by jesserockz 2023/06/23
+"""
+
+from os.path import join
+import glob
+import sys
+
+import subprocess
+
+# pylint: disable=E0602
+Import("env")  # noqa
+
+from SCons.Script import ARGUMENTS
+
+
+platform = env.PioPlatform()
+PROJ_SRC = env["PROJECT_SRC_DIR"]
+PIO_FILES = glob.glob(join(PROJ_SRC, "**", "*.pio"), recursive=True)
+
+verbose = bool(int(ARGUMENTS.get("PIOVERBOSE", "0")))
+
+
+if PIO_FILES:
+    if verbose:
+        print("==============================================")
+        print("PIO ASSEMBLY COMPILER")
+    try:
+        PIOASM_DIR = platform.get_package_dir("tool-pioasm-rp2040-earlephilhower")
+    except:
+        print("tool-pioasm-rp2040-earlephilhower not supported on your system!")
+        sys.exit()
+
+    PIOASM_EXE = join(PIOASM_DIR, "pioasm")
+    if verbose:
+        print("PIO files found:")
+    for filename in PIO_FILES:
+        if verbose:
+            print(f"  {filename}")
+        subprocess.run([PIOASM_EXE, "-o", "c-sdk", filename, "{filename}.h"])
+    if verbose:
+        print("==============================================")

--- a/esphome/components/rp2040_pio/__init__.py
+++ b/esphome/components/rp2040_pio/__init__.py
@@ -1,0 +1,40 @@
+import platform
+
+import esphome.codegen as cg
+
+
+DEPENDENCIES = ["rp2040"]
+
+
+PIOASM_REPO_VERSION = "1.5.0-b"
+PIOASM_REPO_BASE = f"https://github.com/earlephilhower/pico-quick-toolchain/releases/download/{PIOASM_REPO_VERSION}"
+PIOASM_VERSION = "pioasm-2e6142b.230216"
+PIOASM_DOWNLOADS = {
+    "linux": {
+        "aarch64": f"aarch64-linux-gnu.{PIOASM_VERSION}.tar.gz",
+        "armv7l": f"arm-linux-gnueabihf.{PIOASM_VERSION}.tar.gz",
+        "x86_64": f"x86_64-linux-gnu.{PIOASM_VERSION}.tar.gz",
+    },
+    "windows": {
+        "amd64": f"x86_64-w64-mingw32.{PIOASM_VERSION}.zip",
+    },
+    "darwin": {
+        "x86_64": f"x86_64-apple-darwin14.{PIOASM_VERSION}.tar.gz",
+        "arm64": f"x86_64-apple-darwin14.{PIOASM_VERSION}.tar.gz",
+    },
+}
+
+
+async def to_code(config):
+    # cg.add_platformio_option(
+    #     "platform_packages",
+    #     [
+    #         "earlephilhower/tool-pioasm-rp2040-earlephilhower",
+    #     ],
+    # )
+    file = PIOASM_DOWNLOADS[platform.system().lower()][platform.machine().lower()]
+    cg.add_platformio_option(
+        "platform_packages",
+        [f"earlephilhower/tool-pioasm-rp2040-earlephilhower@{PIOASM_REPO_BASE}/{file}"],
+    )
+    cg.add_platformio_option("extra_scripts", ["pre:build_pio.py"])

--- a/esphome/components/rp2040_pio_led_strip/light.py
+++ b/esphome/components/rp2040_pio_led_strip/light.py
@@ -127,6 +127,7 @@ def time_to_cycles(time_us):
 
 CONF_PIO = "pio"
 
+AUTO_LOAD = ["rp2040_pio"]
 CODEOWNERS = ["@Papa-DMan"]
 DEPENDENCIES = ["rp2040"]
 
@@ -265,9 +266,3 @@ async def to_code(config):
                 time_to_cycles(config[CONF_BIT1_LOW]),
             ),
         )
-    cg.add_platformio_option(
-        "platform_packages",
-        [
-            "earlephilhower/tool-pioasm-rp2040-earlephilhower",
-        ],
-    )

